File: codigo/python-sim/kinematics.py
import math

POLE_POSITIONS = [
    [-22.5,-21.0, 45.0], # 0
    [-21.0,-22.5, 45.0], # 1
    [ 21.0,-22.5, 45.0], # 2
    [ 22.5,-21.0, 45.0], # 3
    [ 22.5, 21.0, 45.0], # 4
    [ 21.0, 22.5, 45.0], # 5
    [-21.0, 22.5, 45.0], # 6
    [-22.5, 21.0, 45.0]  # 7
]

ANCHOR_POSITIONS = [
    [-5.0,-4.0, 0.0], # 0 (Corner 1, Anclaje A)
    [-4.0,-5.0, 0.0], # 1 (Corner 1, Anclaje B)
    [ 4.0,-5.0, 0.0], # 2 (Corner 2, Anclaje A)
    [ 5.0,-4.0, 0.0], # 3 (Corner 2, Anclaje B)
    [ 5.0, 4.0, 0.0], # 4 (Corner 3, Anclaje A)
    [ 4.0, 5.0, 0.0], # 5 (Corner 3, Anclaje B)
    [-4.0, 5.0, 0.0], # 6 (Corner 4, Anclaje A)
    [-5.0, 4.0, 0.0]  # 7 (Corner 4, Anclaje B)
]

def calculate_inverse_kinematics(x, y, z, roll, pitch, yaw):
    """
    Calcula la longitud teórica de cada uno de los 8 cables.
    x, y, z en cm. roll, pitch, yaw en radianes.
    Retorna una lista de 8 floats con las longitudes en cm.
    """
    cr, sr = math.cos(roll), math.sin(roll)
    cp, sp = math.cos(pitch), math.sin(pitch)
    cy, sy = math.cos(yaw), math.sin(yaw)

    r11 = cy * cp
    r12 = cy * sp * sr - sy * cr
    r13 = cy * sp * cr + sy * sr

    r21 = sy * cp
    r22 = sy * sp * sr + cy * cr
    r23 = sy * sp * cr - cy * sr

    r31 = -sp
    r32 = cp * sr
    r33 = cp * cr

    lengths = []
    for i in range(8):
        ax = ANCHOR_POSITIONS[i][0]
        ay = ANCHOR_POSITIONS[i][1]
        az = ANCHOR_POSITIONS[i][2]

        qx = x + (r11 * ax + r12 * ay + r13 * az)
        qy = y + (r21 * ax + r22 * ay + r23 * az)
        qz = z + (r31 * ax + r32 * ay + r33 * az)

        dx = POLE_POSITIONS[i][0] - qx
        dy = POLE_POSITIONS[i][1] - qy
        dz = POLE_POSITIONS[i][2] - qz

        lengths.append(math.sqrt(dx**2 + dy**2 + dz**2))

    return lengths

File: codigo/firmware/src/Kinematics.cpp
#include "Kinematics.h"
#include <math.h>

void calculateInverseKinematics(float x, float y, float z, float roll, float pitch, float yaw, float output_lengths[8]) {
    // Calcular cosenos y senos de los ángulos de rotación
    float cr = cos(roll);
    float sr = sin(roll);
    float cp = cos(pitch);
    float sp = sin(pitch);
    float cy = cos(yaw);
    float sy = sin(yaw);

    // Matriz de rotación R_zyx (Roll -> X, Pitch -> Y, Yaw -> Z)
    // R = Rz(yaw) * Ry(pitch) * Rx(roll)
    float r11 = cy * cp;
    float r12 = cy * sp * sr - sy * cr;
    float r13 = cy * sp * cr + sy * sr;

    float r21 = sy * cp;
    float r22 = sy * sp * sr + cy * cr;
    float r23 = sy * sp * cr - cy * sr;

    float r31 = -sp;
    float r32 = cp * sr;
    float r33 = cp * cr;

    // Calcular la longitud para cada uno de los 8 cables
    for (int i = 0; i < 8; i++) {
        // Posición del anclaje de la plataforma respecto a su centro
        float ax = ANCHOR_POSITIONS[i].x;
        float ay = ANCHOR_POSITIONS[i].y;
        float az = ANCHOR_POSITIONS[i].z;

        // Rotar y trasladar el anclaje de la plataforma al espacio del mundo
        float qx = x + (r11 * ax + r12 * ay + r13 * az);
        float qy = y + (r21 * ax + r22 * ay + r23 * az);
        float qz = z + (r31 * ax + r32 * ay + r33 * az);

        // Vector del cable: de la polea fija B_i al anclaje móvil Q_i
        float dx = POLE_POSITIONS[i].x - qx;
        float dy = POLE_POSITIONS[i].y - qy;
        float dz = POLE_POSITIONS[i].z - qz;

        // Longitud del cable (norma L2 del vector)
        output_lengths[i] = sqrt(dx * dx + dy * dy + dz * dz);
    }
}
